### ISSUE TITLE
New maintainer?

### ISSUE THREAD

**Issue by aleeusgr** [walkie/Hagl#8]
hi!
It seems like a good repo without maintenance. I am Haskell newbie looking for a project to adopt.
I have forked the project and working to make it build. Does it make sense to change the code to run with latest versions of packages?
I changed .cabal to have >= insted of == and I get this error:
```
[ 6 of 23] Compiling Hagl.Exec        ( Hagl/Exec.hs, interpreted )

Hagl/Exec.hs:261:21: error:
    • Could not deduce (MonadFail (ExecM g))
        arising from a use of ‘processLocation’
      from the context: (Game g, Eq (Move g))
        bound by the type signature for:
                   step :: forall g. (Game g, Eq (Move g)) => ExecM g (Maybe Payoff)
        at Hagl/Exec.hs:260:1-55
    • In the second argument of ‘(>>=)’, namely ‘processLocation’
      In the expression: location >>= processLocation
      In an equation for ‘step’:
          step
            = location >>= processLocation
            where
                processLocation l
                  = case treeAction l of
                      Decision i -> decide i >>= performMove l
                      Chance d -> fromDist d >>= performMove l
                      Payoff p -> givePayoff p
                performMove l m
                  = do e <- getExec
                       ....
                inc (Decision p) ns = setForPlayer p (forPlayer p ns + 1) ns
                inc _ ns = ns
                decide i
                  = do p <- liftM (forPlayer i) players
                       ....
                ....
    |
261 | step = location >>= processLocation
    |                     ^^^^^^^^^^^^^^^

Hagl/Exec.hs:318:10: error:
    • No instance for (Applicative (ExecM g))
        arising from the superclasses of an instance declaration
    • In the instance declaration for ‘Monad (ExecM g)’
    |
318 | instance Monad (ExecM g) where
    |          ^^^^^^^^^^^^^^^

Hagl/Exec.hs:321:10: error:
    • No instance for (Applicative (StratM s g))
        arising from the superclasses of an instance declaration
    • In the instance declaration for ‘Monad (StratM s g)’
    |
321 | instance Monad (StratM s g) where
    |          ^^^^^^^^^^^^^^^^^^
Failed, five modules loaded.
```
Any suggestion on how I should proceed?
thanks

**Comment by walkie** [comment]
Thanks for your interest! Cleaning up and refactoring this library has been on my to-do list for over a decade now... :grimacing:

Just pushed an update that gets Hagl building again with modern GHC via stack.

I actually did most of the work on this back in 2019, when I started some refactoring work that I didn't finish and never pushed.

The code still isn't great. There are a ton of warnings, bad error handling, and I think some of the types can be simplified. However, it at least now builds! Hopefully that gives you a better starting point for your own experiments.
